File: brainfuck.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"io/ioutil"
)

var (
	code    string
	tape    []uint8
	tapePos int
	codePos int
	reader *bufio.Reader
)

func init() {
	reader = bufio.NewReader(os.Stdin)

	if len(os.Args) > 1 {
		bs, err := ioutil.ReadFile(os.Args[1])

		if err != nil {
			fmt.Println("error while reading the code from a file")
			os.Exit(1)
		}

		code = string(bs)
	} else {
		bs, err := ioutil.ReadAll(os.Stdin)

		if err != nil {
			fmt.Println("error while reading the code from stdin")
			os.Exit(1)
		}

		code = string(bs)
	}
}


func main() {
	run(false)
}


func run(skip bool) bool {
	for tapePos >= 0 && codePos < len(code) {
		if tapePos >= len(tape) {
			tape = append(tape, 0)
		}

		if code[codePos] == '[' {
			codePos++
			oldPos := codePos
			for run(tape[tapePos] == 0) {
				codePos = oldPos
			}
		} else if code[codePos] == ']' {
			return tape[tapePos] != 0

		} else if !skip {
			switch code[codePos] {
			case '+':
				tape[tapePos]++
			case '-':
				tape[tapePos]--
			case '>':
				tapePos++
			case '<':
				tapePos--
			case '.':
				fmt.Print(string(tape[tapePos]))
			case ',':

			}

		}

		codePos++
	}

	return false

}

File: brainfuck.java
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;

public class brainfuck {

	static int codePos = 0, tapePos = 0;

	static ArrayList<Character> tape = new ArrayList<Character>(300);

	static String code;

	public static void main(String[] args) {
		if (args.length > 0) {

			byte[] bytes = null;

			try {
				bytes = Files.readAllBytes(Paths.get(args[0]));
			} catch (IOException e) {
				e.printStackTrace();
			}
			code = new String(bytes, StandardCharsets.UTF_8);
		} else {
			System.out.println("error: reading code from stdin not implemented yet");
			System.exit(1);
		}

		run(false);
	}

	public static boolean run(boolean skip) {
		while (codePos < code.length()) {
			if (tapePos >= tape.size())
				tape.add('\0');

			if (code.charAt(codePos) == '[') {
				++codePos;
				int oldPos = codePos;
				while (run(tape.get(tapePos) == '\0'))
					codePos = oldPos;
			} else if (code.charAt(codePos) == ']') {
				return tape.get(tapePos) != '\0';
			} else if (!skip) {
				switch (code.charAt(codePos)) {
				case '+':
					tape.set(tapePos, (char) (tape.get(tapePos) + 1));
					break;

				case '-':
					tape.set(tapePos, (char) (tape.get(tapePos) - 1));
					break;

				case '>':
					++tapePos;
					break;

				case '<':
					--tapePos;
					break;

				case '.':
					System.out.print(tape.get(tapePos));
					break;

				case ',':
					try {
						tape.set(tapePos, (char) System.in.read());
					} catch (IOException e) {
						e.printStackTrace();
					}
					break;
				}
			}

			++codePos;
		}

		return false;

	}
}
